controllers/machineset_controller_test.go

```diff
@@ -709,3 +737,132 @@ func newMachineSet(name, cluster string) *clusterv1.MachineSet {
 		},
 	}
 }
+
+func TestMachineSetReconcile_MachinesCreatedConditionFalseOnBadInfraRef(t *testing.T) {
+	g := NewWithT(t)
+	replicas := int32(1)
+	version := "v1.21.0"
+	cluster := &clusterv1.Cluster{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "foo",
+			Namespace: metav1.NamespaceDefault,
+		},
+	}
+
+	ms := &clusterv1.MachineSet{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "ms-foo",
+			Namespace: metav1.NamespaceDefault,
+			Labels: map[string]string{
+				clusterv1.ClusterLabelName: cluster.Name,
+			},
+		},
+		Spec: clusterv1.MachineSetSpec{
+			ClusterName: cluster.ObjectMeta.Name,
+			Replicas:    &replicas,
+			Template: clusterv1.MachineTemplateSpec{
+				ObjectMeta: clusterv1.ObjectMeta{
+					Labels: map[string]string{
+						clusterv1.ClusterLabelName: cluster.Name,
+					},
+				},
+				Spec: clusterv1.MachineSpec{
+					InfrastructureRef: corev1.ObjectReference{
+						Kind:       testtypes.GenericInfrastructureMachineTemplateCRD.Kind,
+						APIVersion: testtypes.GenericInfrastructureMachineTemplateCRD.APIVersion,
+						// Try to break Infra Cloning
+						Name:      "something_invalid",
+						Namespace: cluster.Namespace,
+					},
+					Version: &version,
+				},
+			},
+			Selector: metav1.LabelSelector{
+				MatchLabels: map[string]string{
+					clusterv1.ClusterLabelName: cluster.Name,
+				},
+			},
+		},
+	}
+
+	key := util.ObjectKey(ms)
+	request := reconcile.Request{
+		NamespacedName: key,
+	}
+	fakeClient := fake.NewClientBuilder().WithObjects(cluster, ms, testtypes.GenericInfrastructureMachineTemplateCRD.DeepCopy()).Build()
+
+	msr := &MachineSetReconciler{
+		Client:   fakeClient,
+		recorder: record.NewFakeRecorder(32),
+	}
+	_, err := msr.Reconcile(ctx, request)
+	g.Expect(err).To(HaveOccurred())
+	g.Expect(fakeClient.Get(ctx, key, ms)).To(Succeed())
+	gotCond := conditions.Get(ms, clusterv1.MachinesCreatedCondition)
+	g.Expect(gotCond).ToNot(BeNil())
+	g.Expect(gotCond.Status).To(Equal(corev1.ConditionFalse))
+	g.Expect(gotCond.Reason).To(Equal(clusterv1.InfrastructureTemplateCloningFailedReason))
+}
+
+func TestMachineSetReconciler_updateStatusResizedCondition(t *testing.T) {
+	// controller := true
+	cluster := &clusterv1.Cluster{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "foo",
+			Namespace: metav1.NamespaceDefault,
+		},
+	}
+
+	testCases := []struct {
+		name                           string
+		machineSet                     *clusterv1.MachineSet
+		machines                       []*clusterv1.Machine
+		expectedResizedConditionStatus bool
```

Comment:
Looks like this field is currently not used, but can you also add one test case where the ResizeCondition is true, please?

Reply:
I will remove this variable. 
The ResizedCondition=true is already tested in [line 299](https://github.com/kubernetes-sigs/cluster-api/pull/5056/files#diff-5c7510547f6aa537f5810290c80ed60160bfccfb72a3cb13b5067e7e5a5956a5R299).

Alslo, ResizedCondition depends on a workload cluster with Ready Nodes. This complicates this unit test quite a bit.